### ISSUE TITLE
Permission system

### ISSUE THREAD

**Issue by Snowiiii** [Pumpkin-MC/Pumpkin#35]
We want to have an good Permission system using SQL Databases

**Comment by zp3x** [comment]
I can do that.

**Comment by Snowiiii** [comment]
> I can do that.

great, go ahead

**Comment by Erb3** [comment]
I think it makes sense to use SQLite, or a similar file-based database.

**Comment by Snowiiii** [comment]
> I think it makes sense to use SQLite, or a similar file-based database.

Yep, I already had that in mind

**Comment by Bryntet** [comment]
Why would we want a database for permissions, and not just store it on the player? Or are we thinking its stored on both? Like its loaded from a db when player joins?

**Comment by Wundero** [comment]
> Why would we want a database for permissions, and not just store it on the player? Or are we thinking its stored on both? Like its loaded from a db when player joins?

Storing permissions in a database allows the server to have a lower memory overhead, especially for servers which have large numbers of permissions / online players. Database performance is quite fast and highly optimized for memory and cpu usage, and queries against permissions are relatively simple to index, so for most practical purposes, running something like SQLite in the background will be better for performance and maintainability than storing a bunch of strings on the player's struct in memory. No sense reinventing the wheel here IMO. 


> I think it makes sense to use SQLite, or a similar file-based database.

I think it would be nice to be able to plug+play the database in the long run, so users who want databases running on separate servers can provide a connection URL and swap from SQLite to MySQL, PostgreSQL, or some other implementation or machine, but this would be a lower priority than just implementing it to begin with on SQLite.

**Comment by lokka30** [comment]
It would be very interesting if somehow LuckPerms can be ported to work far down the line. Potentially that plugin would be a great reference on what features would be useful here - particularly, design choices made in its [API](https://luckperms.net/wiki/Developer-API-Usage).

**Comment by zp3x** [comment]
I will try to implement sources like LuckPerms as well.

**Comment by vyPal** [comment]
_(Copied from [discord](https://discord.com/channels/1268592337445978193/1268611318617866261/1337474809226858527))_
ok so I did a little bit of digging and apparently bukkit's plugin system only temporarily stores which permissions exist and which player has which permissions. It also allows plugins to query the permissions of a player, modify the permissions, etc. It is the job of a plugin like luckperms to store the permissions somewhere (like a db) and tell bukkit who has which permissions. Now the question is if we want to take a similar approach.

**Comment by vyPal** [comment]
> Wouldn't a permission system as a plugin (just like LuckPerms) be better

Well that's kinda the problem, luckperms is **not** a permission system. It is only responsible for storing which player has which permissions (and of course it has some other features, but those aren't a requirement). The server itself (Bukkit, spigot, etc.) contains the logic for validating permissions, and the APIs to let plugins set and query permissions. 

> to keep core as close to vanilla as possible?

I also see vanilla parity as an important thing for this project, however we are already quite far off vanilla parity just by having a plugin system. Now in vanilla there is essentially just the op "permission" which can have different levels to allow different actions to be performed by server operators. If we introduced a node-based permissions system (`pumpkin.command.setblock` for example) into the core of Pumpkin, it wouldn't change anything for users who don't use any plugins. This is because in Bukkit's node based system, there is the option to set a "default" parameter, this determines which players receive the given role unless a permission management plugin (like luckperms) says otherwise. 

So in short, a node-based permission system (inspired by Bukkit or spigot) would appear to work the same as vanilla for anyone not using plugins, but for servers using plugins, it would allow them to coordinate permissions between each other without having to all depend on an extra plugin

**Comment by Q2297045667** [comment]
> Wouldn't a permission system as a plugin (just like LuckPerms) be better to keep core as close to vanilla as possible?

I think with the current capabilities of the Minecraft community,it’s unlikely that anyone would develop a plugin like LuckPerms,or we could just build in a simple permission system.

**Comment by Snowiiii** [comment]
I think the current String based "Permission system" is enough and looking back using an SQL database would be way too overengineered, If there is a strong need, i'm fine with reopening the discussion but as it is i think we should leave complex non vanilla permission that many may not need up to plugins
